File: apps/merchant-terminal-android/app/src/main/java/com/yunqiao/life/merchantterminal/printing/receipt/ReceiptDocumentV1.kt
package com.yunqiao.life.merchantterminal.printing.receipt

import org.json.JSONObject
import java.time.Instant

enum class ReceiptType { ORDER_CUSTOMER, TABLE_BILL }

data class ReceiptMerchant(
    val id: String,
    val name: String,
    val address: String?,
    val phone: String?,
)

data class ReceiptOrder(
    val id: String,
    val orderNo: String,
    val orderType: String,
    val tableName: String?,
    val guestCount: Int?,
    val createdAt: String,
    val completedAt: String?,
)

data class ReceiptTableSession(
    val id: String,
    val sessionNo: String,
    val tableName: String,
    val openedAt: String,
    val closedAt: String?,
    val orderNos: List<String>,
)

data class ReceiptItem(
    val name: String,
    val nameVi: String?,
    val nameEn: String?,
    val quantity: Int,
    val unitPrice: Long,
    val lineTotal: Long,
    val specification: String?,
    val note: String?,
)

data class ReceiptTotals(
    val subtotal: Long,
    val discount: Long?,
    val serviceFee: Long?,
    val total: Long,
    val currency: String,
)

data class ReceiptDocumentV1(
    val receiptType: ReceiptType,
    val generatedAt: String,
    val merchant: ReceiptMerchant,
    val order: ReceiptOrder?,
    val tableSession: ReceiptTableSession?,
    val items: List<ReceiptItem>,
    val totals: ReceiptTotals,
    val note: String?,
    val verificationCode: String?,
)

/** Strict parser: rejects unknown/oversized fields and never accepts HTML or printer bytes. */
object ReceiptDocumentParser {
    fun parse(json: String): ReceiptDocumentV1 {
        require(json.length in 2..MAX_JSON_CHARS) { "Receipt snapshot size is invalid." }
        val root = JSONObject(json)
        root.requireOnly(
            "schemaVersion", "receiptType", "generatedAt", "merchant", "order",
            "tableSession", "items", "totals", "note", "verificationCode",
        )
        require(root.optInt("schemaVersion", -1) == 1) { "Unsupported receipt schema." }
        val receiptType = enumValueOf<ReceiptType>(root.requiredText("receiptType", 32))
        val generatedAt = root.requiredInstant("generatedAt")
        val merchantObject = root.requiredObject("merchant").also {
            it.requireOnly("id", "name", "address", "phone")
        }
        val merchant = ReceiptMerchant(
            id = merchantObject.requiredNumericId("id"),
            name = merchantObject.requiredText("name", 120),
            address = merchantObject.optionalText("address", 300),
            phone = merchantObject.optionalText("phone", 32),
        )
        val order = root.optJSONObject("order")?.let(::parseOrder)
        val tableSession = root.optJSONObject("tableSession")?.let(::parseTableSession)
        require(
            (receiptType == ReceiptType.ORDER_CUSTOMER && order != null && tableSession == null) ||
                (receiptType == ReceiptType.TABLE_BILL && tableSession != null && order == null),
        ) { "Receipt context does not match receipt type." }
        val itemArray = root.optJSONArray("items") ?: error("Receipt items are missing.")
        require(itemArray.length() in 1..500) { "Receipt item count is invalid." }
        val items = (0 until itemArray.length()).map { index ->
            val item = itemArray.optJSONObject(index) ?: error("Receipt item is invalid.")
            item.requireOnly(
                "name", "nameVi", "nameEn", "quantity", "unitPrice", "lineTotal",
                "specification", "note",
            )
            ReceiptItem(
                name = item.requiredText("name", 120),
                nameVi = item.optionalText("nameVi", 120),
                nameEn = item.optionalText("nameEn", 120),
                quantity = item.requiredPositiveInt("quantity"),
                unitPrice = item.requiredNonNegativeLong("unitPrice"),
                lineTotal = item.requiredNonNegativeLong("lineTotal"),
                specification = item.optionalText("specification", 120),
                note = item.optionalText("note", 200),
            )
        }
        val totalsObject = root.requiredObject("totals").also {
            it.requireOnly("subtotal", "discount", "serviceFee", "total", "currency")
        }
        val totals = ReceiptTotals(
            subtotal = totalsObject.requiredNonNegativeLong("subtotal"),
            discount = totalsObject.optionalNonNegativeLong("discount"),
            serviceFee = totalsObject.optionalNonNegativeLong("serviceFee"),
            total = totalsObject.requiredNonNegativeLong("total"),
            currency = totalsObject.requiredText("currency", 8).also { require(it == "VND") },
        )
        return ReceiptDocumentV1(
            receiptType = receiptType,
            generatedAt = generatedAt,
            merchant = merchant,
            order = order,
            tableSession = tableSession,
            items = items,
            totals = totals,
            note = root.optionalText("note", 500),
            verificationCode = root.optionalText("verificationCode", 128),
        )
    }

    private fun parseOrder(value: JSONObject): ReceiptOrder {
        value.requireOnly(
            "id", "orderNo", "orderType", "tableName", "guestCount", "createdAt", "completedAt",
        )
        return ReceiptOrder(
            id = value.requiredNumericId("id"),
            orderNo = value.requiredText("orderNo", 32),
            orderType = value.requiredText("orderType", 32),
            tableName = value.optionalText("tableName", 64),
            guestCount = value.optionalNonNegativeInt("guestCount"),
            createdAt = value.requiredInstant("createdAt"),
            completedAt = value.optionalInstant("completedAt"),
        )
    }

    private fun parseTableSession(value: JSONObject): ReceiptTableSession {
        value.requireOnly("id", "sessionNo", "tableName", "openedAt", "closedAt", "orderNos")
        val orderNosArray = value.optJSONArray("orderNos") ?: error("orderNos are missing.")
        require(orderNosArray.length() <= 1_000)
        return ReceiptTableSession(
            id = value.requiredNumericId("id"),
            sessionNo = value.requiredText("sessionNo", 32),
            tableName = value.requiredText("tableName", 64),
            openedAt = value.requiredInstant("openedAt"),
            closedAt = value.optionalInstant("closedAt"),
            orderNos = (0 until orderNosArray.length()).map {
                orderNosArray.optString(it).also { orderNo -> require(orderNo.length in 1..32) }
            },
        )
    }

    private fun JSONObject.requireOnly(vararg allowed: String) {
        val allowedSet = allowed.toSet()
        require(keys().asSequence().all { it in allowedSet }) { "Receipt contains unsupported fields." }
    }

    private fun JSONObject.requiredObject(key: String): JSONObject = optJSONObject(key)
        ?: error("Missing receipt object: $key")

    private fun JSONObject.requiredText(key: String, max: Int): String = optString(key)
        .takeIf { it.length in 1..max && !it.contains('\u0000') }
        ?: error("Invalid receipt text: $key")

    private fun JSONObject.optionalText(key: String, max: Int): String? = if (isNull(key)) null else
        optString(key).takeIf { it.length <= max && !it.contains('\u0000') }
            ?: error("Invalid receipt text: $key")

    private fun JSONObject.requiredNumericId(key: String): String = requiredText(key, 40)
        .also { require(NUMERIC_ID.matches(it)) }

    private fun JSONObject.requiredInstant(key: String): String = requiredText(key, 40)
        .also { Instant.parse(it) }

    private fun JSONObject.optionalInstant(key: String): String? = optionalText(key, 40)
        ?.also { Instant.parse(it) }

    private fun JSONObject.requiredPositiveInt(key: String): Int = optInt(key, -1)
        .also { require(it > 0) }

    private fun JSONObject.optionalNonNegativeInt(key: String): Int? = if (isNull(key)) null else
        optInt(key, -1).also { require(it >= 0) }

    private fun JSONObject.requiredNonNegativeLong(key: String): Long = optLong(key, -1)
        .also { require(it >= 0) }

    private fun JSONObject.optionalNonNegativeLong(key: String): Long? = if (isNull(key)) null else
        optLong(key, -1).also { require(it >= 0) }

    private const val MAX_JSON_CHARS = 512_000
    private val NUMERIC_ID = Regex("^[0-9]+$")
}

File: clients/windows/YunQiao.Cashier/src/YunQiao.Cashier.Core/Protocol/ReceiptDocumentV1.cs
using System.Text.Json;

namespace YunQiao.Cashier.Core.Protocol;

public enum ReceiptType { ORDER_CUSTOMER, TABLE_BILL }

public sealed record ReceiptMerchant(string Id, string Name, string? NameVi, string? Address, string? Phone);
public sealed record ReceiptOrder(string Id, string OrderNo, string OrderType, string? TableName, int? GuestCount, DateTimeOffset CreatedAt, DateTimeOffset? CompletedAt);
public sealed record ReceiptTableSession(string Id, string SessionNo, string TableName, DateTimeOffset OpenedAt, DateTimeOffset? ClosedAt, IReadOnlyList<string> OrderNos);
public sealed record ReceiptItem(string Name, string? NameVi, string? NameEn, int Quantity, long UnitPrice, long LineTotal, string? Specification, string? Note);
public sealed record ReceiptTotals(long Subtotal, long? Discount, long? OriginalAmount, long? RoundingAmount, long? ReceivedAmount, long? ServiceFee, long Total, string Currency);
public sealed record ReceiptFooter(string Zh, string Vi);
public sealed record ReceiptDocumentV1(
    ReceiptType ReceiptType,
    DateTimeOffset GeneratedAt,
    ReceiptMerchant Merchant,
    ReceiptOrder? Order,
    ReceiptTableSession? TableSession,
    IReadOnlyList<ReceiptItem> Items,
    ReceiptTotals Totals,
    string? Note,
    string? VerificationCode,
    ReceiptFooter? Footer);

/// <summary>Strict port of Android ReceiptDocumentParser for active schema 1 jobs.</summary>
public static class ReceiptDocumentV1Parser
{
    public static ReceiptDocumentV1 Parse(string json)
    {
        if (json.Length is < 2 or > PrintDocumentParser.MaxJsonChars) throw new ReceiptSchemaException("Receipt snapshot size is invalid.");
        using var parsed = JsonDocument.Parse(json, new JsonDocumentOptions { MaxDepth = 64 });
        var root = parsed.RootElement;
        RequireObject(root, "$");
        RequireOnly(root, "$", "schemaVersion", "receiptType", "generatedAt", "merchant", "order", "tableSession", "items", "totals", "note", "verificationCode", "footer");
        if (RequiredInt(root, "schemaVersion", 1, 1) != 1) throw new ReceiptSchemaException("Unsupported receipt schema.");
        var receiptType = RequiredEnum<ReceiptType>(root, "receiptType", 32);
        var merchantJson = RequiredObject(root, "merchant");
        RequireOnly(merchantJson, "$.merchant", "id", "name", "nameVi", "address", "phone");
        var merchant = new ReceiptMerchant(
            RequiredNumericId(merchantJson, "id"),
            RequiredText(merchantJson, "name", 120),
            OptionalText(merchantJson, "nameVi", 120),
            OptionalText(merchantJson, "address", 300),
            OptionalText(merchantJson, "phone", 32));

        var order = OptionalObject(root, "order") is { } orderJson ? ParseOrder(orderJson) : null;
        var table = OptionalObject(root, "tableSession") is { } tableJson ? ParseTable(tableJson) : null;
        if ((receiptType == ReceiptType.ORDER_CUSTOMER && (order is null || table is not null))
            || (receiptType == ReceiptType.TABLE_BILL && (table is null || order is not null)))
            throw new ReceiptSchemaException("Receipt context does not match receipt type.");

        if (!root.TryGetProperty("items", out var itemsJson) || itemsJson.ValueKind != JsonValueKind.Array || itemsJson.GetArrayLength() is < 1 or > 500)
            throw new ReceiptSchemaException("Receipt item count is invalid.");
        var items = new List<ReceiptItem>(itemsJson.GetArrayLength());
        var index = 0;
        foreach (var item in itemsJson.EnumerateArray())
        {
            RequireObject(item, $"$.items[{index}]");
            RequireOnly(item, $"$.items[{index}]", "name", "nameVi", "nameEn", "quantity", "unitPrice", "lineTotal", "specification", "note");
            items.Add(new ReceiptItem(
                RequiredText(item, "name", 120), OptionalText(item, "nameVi", 120), OptionalText(item, "nameEn", 120),
                RequiredInt(item, "quantity", 1, int.MaxValue), RequiredLong(item, "unitPrice"), RequiredLong(item, "lineTotal"),
                OptionalText(item, "specification", 120), OptionalText(item, "note", 200)));
            index++;
        }

        var totalsJson = RequiredObject(root, "totals");
        RequireOnly(totalsJson, "$.totals", "subtotal", "discount", "originalAmount", "roundingAmount", "receivedAmount", "serviceFee", "total", "currency");
        var totals = new ReceiptTotals(
            RequiredLong(totalsJson, "subtotal"), OptionalLong(totalsJson, "discount"), OptionalLong(totalsJson, "originalAmount"),
            OptionalLong(totalsJson, "roundingAmount"), OptionalLong(totalsJson, "receivedAmount"), OptionalLong(totalsJson, "serviceFee"),
            RequiredLong(totalsJson, "total"), RequiredText(totalsJson, "currency", 8));
        if (totals.Currency != "VND") throw new ReceiptSchemaException("Unsupported receipt currency.");

        ReceiptFooter? footer = null;
        if (OptionalObject(root, "footer") is { } footerJson)
        {
            RequireOnly(footerJson, "$.footer", "zh", "vi");
            footer = new ReceiptFooter(RequiredText(footerJson, "zh", 60), RequiredText(footerJson, "vi", 60));
        }
        return new ReceiptDocumentV1(
            receiptType, RequiredInstant(root, "generatedAt"), merchant, order, table, items, totals,
            OptionalText(root, "note", 500), OptionalText(root, "verificationCode", 128), footer);
    }

    private static ReceiptOrder ParseOrder(JsonElement value)
    {
        RequireOnly(value, "$.order", "id", "orderNo", "orderType", "tableName", "guestCount", "createdAt", "completedAt");
        return new ReceiptOrder(
            RequiredNumericId(value, "id"), RequiredText(value, "orderNo", 32), RequiredText(value, "orderType", 32),
            OptionalText(value, "tableName", 64), OptionalInt(value, "guestCount"), RequiredInstant(value, "createdAt"), OptionalInstant(value, "completedAt"));
    }

    private static ReceiptTableSession ParseTable(JsonElement value)
    {
        RequireOnly(value, "$.tableSession", "id", "sessionNo", "tableName", "openedAt", "closedAt", "orderNos");
        if (!value.TryGetProperty("orderNos", out var orderNosJson) || orderNosJson.ValueKind != JsonValueKind.Array || orderNosJson.GetArrayLength() > 1_000)
            throw new ReceiptSchemaException("orderNos are invalid.");
        var orderNos = orderNosJson.EnumerateArray().Select(item =>
        {
            if (item.ValueKind != JsonValueKind.String || item.GetString() is not { Length: >= 1 and <= 32 } text)
                throw new ReceiptSchemaException("orderNo is invalid.");
            return text;
        }).ToArray();
        return new ReceiptTableSession(
            RequiredNumericId(value, "id"), RequiredText(value, "sessionNo", 32), RequiredText(value, "tableName", 64),
            RequiredInstant(value, "openedAt"), OptionalInstant(value, "closedAt"), orderNos);
    }

    private static void RequireOnly(JsonElement value, string path, params string[] allowed)
    {
        var accepted = allowed.ToHashSet(StringComparer.Ordinal);
        var unsupported = value.EnumerateObject().Select(item => item.Name).Where(item => !accepted.Contains(item)).Order(StringComparer.Ordinal).ToArray();
        if (unsupported.Length > 0) throw new ReceiptSchemaException($"Receipt contains unsupported fields at {path}: {string.Join(',', unsupported)}.");
    }

    private static void RequireObject(JsonElement value, string path)
    {
        if (value.ValueKind != JsonValueKind.Object) throw new ReceiptSchemaException($"Receipt value at {path} must be an object.");
    }

    private static JsonElement RequiredObject(JsonElement owner, string key) =>
        owner.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.Object ? value : throw Invalid(key);

    private static JsonElement? OptionalObject(JsonElement owner, string key) =>
        !owner.TryGetProperty(key, out var value) || value.ValueKind == JsonValueKind.Null ? null
            : value.ValueKind == JsonValueKind.Object ? value : throw Invalid(key);

    private static string RequiredText(JsonElement owner, string key, int max) =>
        OptionalText(owner, key, max) is { Length: > 0 } value ? value : throw Invalid(key);

    private static string? OptionalText(JsonElement owner, string key, int max)
    {
        if (!owner.TryGetProperty(key, out var value) || value.ValueKind == JsonValueKind.Null) return null;
        if (value.ValueKind != JsonValueKind.String || value.GetString() is not { } text || text.Length > max || text.Contains('\0')) throw Invalid(key);
        return text;
    }

    private static string RequiredNumericId(JsonElement owner, string key)
    {
        var value = RequiredText(owner, key, 40);
        if (!value.All(char.IsAsciiDigit)) throw Invalid(key);
        return value;
    }

    private static int RequiredInt(JsonElement owner, string key, int min, int max) =>
        owner.TryGetProperty(key, out var value) && value.TryGetInt32(out var result) && result >= min && result <= max ? result : throw Invalid(key);

    private static int? OptionalInt(JsonElement owner, string key)
    {
        if (!owner.TryGetProperty(key, out var value) || value.ValueKind == JsonValueKind.Null) return null;
        return value.TryGetInt32(out var result) && result >= 0 ? result : throw Invalid(key);
    }

    private static long RequiredLong(JsonElement owner, string key) =>
        owner.TryGetProperty(key, out var value) && value.TryGetInt64(out var result) && result >= 0 ? result : throw Invalid(key);

    private static long? OptionalLong(JsonElement owner, string key)
    {
        if (!owner.TryGetProperty(key, out var value) || value.ValueKind == JsonValueKind.Null) return null;
        return value.TryGetInt64(out var result) && result >= 0 ? result : throw Invalid(key);
    }

    private static DateTimeOffset RequiredInstant(JsonElement owner, string key) =>
        OptionalInstant(owner, key) ?? throw Invalid(key);

    private static DateTimeOffset? OptionalInstant(JsonElement owner, string key)
    {
        var text = OptionalText(owner, key, 40);
        if (text is null) return null;
        return DateTimeOffset.TryParse(text, out var result) ? result : throw Invalid(key);
    }

    private static T RequiredEnum<T>(JsonElement owner, string key, int max) where T : struct, Enum =>
        Enum.TryParse<T>(RequiredText(owner, key, max), false, out var result) && Enum.IsDefined(result) ? result : throw Invalid(key);

    private static ReceiptSchemaException Invalid(string key) => new($"Invalid receipt field: {key}");
}

public sealed class ReceiptSchemaException(string message) : Exception(message);
